src/ucp/core/ucp_listener.c

```diff
@@ -42,77 +60,57 @@ void ucp_listener_schedule_accept_cb(ucp_ep_h ep)
 
 static unsigned ucp_listener_conn_request_progress(void *arg)
 {
-    ucp_listener_accept_t            *accept      = arg;
-    const ucp_wireup_sockaddr_priv_t *client_data = accept->wireup_data;
-    ucp_listener_h                   listener     = accept->listener;
-    ucp_unpacked_address_t           client_address;
-    ucp_ep_params_t                  params;
+    ucp_conn_request_h               conn_request = arg;
+    ucp_listener_h                   listener     = conn_request->listener;
+    const ucp_wireup_client_data_t   *client_data = &conn_request->client_data;
+    ucp_worker_h                     worker;
     ucp_ep_h                         ep;
     ucs_status_t                     status;
 
     ucs_trace_func("listener=%p", listener);
 
-    status = ucp_address_unpack(client_data + 1, &client_address);
-    if (status != UCS_OK) {
-        goto out;
+    if (listener->conn_cb) {
+        listener->conn_cb(conn_request, listener->arg);
+        return 1;
     }
 
-    params.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
-    params.err_mode   = client_data->err_mode;
-
-    if (client_data->is_full_addr) {
-        /* create endpoint to the worker address we got in the private data */
-        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
-                                              &client_address,
-                                              UCP_EP_CREATE_AM_LANE, "listener",
-                                              &ep);
-        if (status == UCS_OK) {
-            ucp_ep_update_dest_ep_ptr(ep, client_data->ep_ptr);
-            ucp_ep_flush_state_reset(ep);
+    worker = listener->wiface.worker;
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(worker);
+    UCS_ASYNC_BLOCK(&worker->async);
+    /* coverity[overrun-buffer-val] */
+    status = ucp_ep_create_accept(worker, client_data, &ep);
+    UCS_ASYNC_UNBLOCK(&worker->async);
+    UCP_WORKER_THREAD_CS_EXIT_CONDITIONAL(worker);
 
-            /* send wireup request message, to connect the client to the server's new endpoint */
+    if (status == UCS_OK) {
+        if (ep->flags & UCP_EP_FLAG_LISTENER) {
+            status = ucp_wireup_send_pre_request(ep);
```

Comment:
are we calling this without lock?

Reply:
ep is created, ucp request for wireup is allocated by malloc (not form worker's pool), why lock is needed here? ep state shouldn't be changed by remote side because it's "pre-request", i missed something?
But `ucp_wireup_send_request` at line 95 should be locked to prevent possible race on `ep->conn_sn`, right?

Reply:
ucp_wireup_send_pre_request() can go to the transport (e.g ud) which is not thread-safe